### ISSUE TITLE
Could the compiler utilize OptimizedClosures.FSharpFunc?

### ISSUE THREAD

**Issue by atlemann** [Microsoft/visualfsharp#6499]
I was doing some benchmarking today and was a bit surprised by some of the results. 

The following function

```fsharp
let bar (f : 'a -> 'a -> 'a) (a:'a array) (b:'a array) =
    Array.map2 f a b
```

is much faster than the next one, even though the previous one allocates a whole new array to store the results in.

```fsharp
let barInPlace (f : 'a -> 'a -> 'a) (a:'a array) (b:'a array) =
    for i in 0..b.Length - 1 do
        b.[i] <- f a.[i] b.[i]
```
In my case, I'd rather mutate one of the arrays instead of allocating a new one, since it can be quite large and `b` would be GC'd if not.

#### Repro steps

https://github.com/atlemann/OptimizedClosures

```bash
dotnet run -c Release -p OptimizedFunc.fsproj
```

#### Expected behavior

`barInPlace` to be faster than `bar`

#### Actual behavior

`bar` is much faster even though it allocates a whole new array

#### Known workarounds

Use OptimizedClosures.FSharpFunc in my own code, or inline all the things.

#### Related information 

.NET Core SDK (reflecting any global.json):
 Version:   2.2.104
 Commit:    73f036d4ac

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  18.04
 OS Platform: Linux
 RID:         ubuntu.18.04-x64

**Comment by cartermp** [comment]
I'm going to assume a good reason for not adapting any user-defined mapping function `mapping` via`OptimizedClosures.FSharpFunc<_, _, _>.Adapt(mapping)` when passed as input, but it could very well be that there isn't one.

To make matters weirder, pasting in your benchmark to [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzgHwFsIATAVwxgAIAxCCKgXgFgAoNqzqygFyuACGUKgApsVEFQDkAqnAB802QqUBKUQJAyqQqAICe6kaG26D6lqy7WqAQSh79AOkICADgCYq42cDYcuXn4hAEkAOwAFDAEwajEJJTlFbRUZI01TB3NRE1kzQyYAm29oKgBLcrCqAAYnJ2AnABkYMIBzHgALOSoARioSCCLirgaAbTKAXSoAHjhvHSdxqbHJ/ytAmD4ysIxt6kEocKiYmHDdsLjxSRTk5Vv0rTysguNHnWeLIetsUortmrqDWabU63T6Ay+wxWU1m8wEi0m/AREzW1iCByO0ViAHk3DwAMIYCC4MiwURXRKpO5qDRvfJGXLvRyfdbFILYMhhMBMKi4nhlQhlABeMBIhOJpJguCcNAAyh0hG4aJywNMZGglBqZPInLYSO4eGJVJDOD9hH8qrV6k0Wu0unNwYNWcNONCZnMOVynOEAG4QADWMBE8KWGuhxvYzqC23O+1CkSxMD54pJsDOe3JCRulPutMyzJydI+hWdaM23hVPL5AuFopTkulcoVUCVKrVAi1HaUOr1BqNJpK5sqAOtwLtYP6Tpd1jdsM9YG9YT9geDyLDyIjQ3RQmVYXCuDKJDimgPbUoTOyr1PrXP9JLLvZPLEnJ0/CSr4AVACAOziADUb5frUADM2ARi6ZrlMOVpAraoIOpOA4zsi7pwihbpAA===) shows a variety of different things these calls get compiled into. Most notable is that the inner function approach gets most optimized.

In general the optimization pass of the compiler is tricky business. But I certainly think we'd welcome a PR that demonstrates what could be done, perhaps having `mapping` adapt in your example.

**Comment by NinoFloris** [comment]
>I'm going to assume a good reason for not adapting any user-defined mapping function mapping via OptimizedClosures.FSharpFunc<_, _, _>.Adapt(mapping) when passed as input, but it could very well be that there isn't one.

@dsyme would it make sense to have the compiler optimize calls when a function is called in a top level loop construct?
There's still a ~2.5x diff on net5.0 in perf due to the type test in InvokeFast that is avoided by adapting. The only downside I see is that it introduces one extra (invisible) alloc for side effecting functions (but those would already alloc per call anyway). 

Adding adapt in inner loops is a bit trickier precisely due to the potential for side effecting functions in perf sensitive loops to change their allocation profile from say 1 to 2 allocs per outer iteration. Unless it's easy to determine when to bail if it cannot be hoisted to the function body?

Could we update the title to reflect the issue (specifically being about Adapt) a bit better?

**Comment by dsyme** [comment]
> @dsyme would it make sense to have the compiler optimize calls when a function is called in a top level loop construct?

Yes, it could in theory emit this.  We don't do any loop-hoisting optimization like this however, and in theory you might also want to apply it to tail-recursive functions.  Tricky though.
